apps/system/js/bluetooth_v2.js

```diff
@@ -0,0 +1,408 @@
+
+  /**
+   * Build-in Bluetooth profiles.
+   *
+   * @public
+   */
+  get Profiles() {
+    return {
+      HFP: 'hfp',   // Hands-Free Profile
+      OPP: 'opp',   // Object Push Profile
+      A2DP: 'a2dp', // A2DP status
+      SCO: 'sco'    // Synchronous Connection-Oriented
+    };
+  },
+
+  /**
+   * Set profile connect state.
+   *
+   * @public
+   * @param {String} profile profile id
+   * @param {Boolean} connected connect status
+   */
+  _setProfileConnected: function bt_setProfileConnected(profile, connected) {
+    var value = this['_' + profile + 'Connected'];
+    if (value !== connected) {
+      this['_' + profile + 'Connected'] = connected;
+
+      // Raise an event for the profile connection changes.
+      window.dispatchEvent(new CustomEvent('bluetoothprofileconnectionchange',
+      {
+        name: profile,
+        connected: connected
+      }));
+      // if (profile === 'opp' && this.transferIcon) {
+      //   this.transferIcon.update();
+      // }
+      // if (profile === 'a2dp' && this.headphoneIcon) {
+      //   this.headphoneIcon.update();
+      // }
+    }
+  },
+
+  /**
+   * Check if bluetooth profile is connected.
+   *
+   * @public
+   * @param {String} profile profile name
+   * @return {Boolean} connected state
+   */
+  isProfileConnected: function bt_isProfileConnected(profile) {
+    var isConnected = this['_' + profile + 'Connected'];
+    if (isConnected === undefined) {
+      return false;
+    } else {
+      return isConnected;
+    }
+  },
+
+  /**
+   * Initialize bluetooth module.
+   *
+   * @public
+   */
+  start: function bt_start() {
+    if (!window.navigator.mozSettings || !window.navigator.mozBluetooth) {
+      return;
+    }
+
+    this._bluetooth = window.navigator.mozBluetooth;
+
+    this._defaultAdapterChangeHandler =
+      this._bluetoothManagerHandler.bind(this, null);
+    this._bluetooth.addEventListener('attributechanged',
+      this._defaultAdapterChangeHandler);
+
+    // clear defaultAdapter and listener once adapter is removed
+    this._bluetooth.addEventListener('adapterremoved',
+      this._cleanupListenerAndAdapters.bind(this));
+
+    window.addEventListener('bluetooth-available', (evt) => {
+      this._debug('adapter is available');
+      this._stateChangeHandler =
+        this._bluetoothAdapterHandler.bind(this);
+      this._adapter.addEventListener('attributechanged',
+        this._stateChangeHandler);
+      this._listenProfileStates();
+    });
+
+    // decouple bluetooth enable/disable function from other system part
+    window.addEventListener('request-enable-bluetooth',
+      this._requestEnableHandler.bind(this));
+    window.addEventListener('request-disable-bluetooth',
+      this._requestDisableHandler.bind(this));
+
+    // Generally bluetooth adapter is already available when start
+    // this module and no related event will be received.
+    // so we try to get the adapter right now
+    this._debug('init bluetooth adapter');
```

Comment:
the init adapter handler is now moving before the other adapter listeners to prevent race condition.